### PR TITLE
remove bare var handling in conditionals

### LINKED ISSUE #39414: Toplevel and Sublevel Variables Treated Differently by Conditionals
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

conditional.py

##### ANSIBLE VERSION

```
ansible 2.6.0 (devel efe7c20100) last updated 2018/04/27 11:30:28 (GMT +100)
  config file = None
  configured module search path = [u'/Users/dennis.conrad/github/ansible/lib/ansible/modules']
  ansible python module location = /Users/dennis.conrad/github/ansible/lib/ansible
  executable location = bin/ansible
  python version = 2.7.14 (default, Nov 20 2017, 11:18:37) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.42)]
```

##### CONFIGURATION

DEFAULT_MODULE_PATH(env: ANSIBLE_LIBRARY) = [u'/Users/dennis.conrad/github/ansible/lib/ansible/modules']

##### OS / ENVIRONMENT

N/A

##### SUMMARY

When checking whether a variable is a boolean or not empty, conditionals behave differently for toplevel variables and sublevel variables; for example, given

```yaml
set_fact:
  toplevel: teststring
  hash:
    var_in_hash: teststring
```

,
```yaml
when: toplevel
```

behaves differently (i. e. breaks), compared to
```yaml
when: hash.var_in_hash
```

, which works as expected.  The former fails with

```json
fatal: [localhost]: FAILED! => {
    "msg": "The conditional check 'toplevel' failed. The error was: error while evaluating conditional (toplevel): 'teststring' is undefined\n\nThe error appears to have been in '/Users/dennis.conrad/github/ansible/testcase.yml': line 23, column 5, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n  - debug:\n    ^ here\n"
}
```

In this case, a variable named `teststring` (the value of the `toplevel` variable) cannot be found, but certain different values for toplevel variables (slashes, dots, spaces...) break the jinja2 parser in other ways.

This is because in `lib/ansible/playbook/conditional.py`, the value of a toplevel variable is first extracted

```python
132         if conditional in all_vars and VALID_VAR_REGEX.match(conditional):
133             conditional = all_vars[conditional]
```

, then presented to the jinja2 parser (instead of the variable name):

```python
192             presented = "{%% if %s %%} True {%% else %%} False {%% endif %%}" % conditional
193             val = templar.template(presented, disable_lookups=disable_lookups).strip()
```

Of course, the value of a toplevel variable doesn't necessarily exist in its own right as a variable and cannot be found (and - if found - could potentially cause other issues), and if the value of the toplevel variable contains any characters that have a special meaning to jinja2 (slashes, dots, spaces...), the jinja2 parser fails as well, for obvious reasons (run the playbook in the [STEPS TO REPRODUCE](#steps-to-reproduce) section for more examples).

Removing the following code from `lib/ansible/playbook/conditional.py`solves the issue (this can, again, be tested with the playbook in the [STEPS TO REPRODUCE](#steps-to-reproduce) section.):

```python
125         # pull the "bare" var out, which allows for nested conditionals
126         # and things like:
127         # - assert:
128         #     that:
129         #     - item
130         #   with_items:
131         #   - 1 == 1
132         if conditional in all_vars and VALID_VAR_REGEX.match(conditional):
133             conditional = all_vars[conditional]
```

The above check appears to have been re-introduced in https://github.com/ansible/ansible/commit/0e834509c85a22e51debf48efe966dea2c9cdaa4#diff-2d3db4a48d45561a69c3b8064cd26649 , when `convert_bare=True` was removed from the `templar.template()` call.

Since using bare variables is not supported in Ansible anymore, it seems to be that above code can (and should) be removed.

##### STEPS TO REPRODUCE

Example playbook which shows the problem:

https://gist.github.com/dennisconrad/1906ff16131d3f9eb2ea4df7391a00b5

##### EXPECTED RESULTS

Toplevel variables and sublevel variables should be handled the same.

##### ACTUAL RESULTS

Toplevel variables are handled differently to sublevel variables; the former break the jinja2 parser, the latter don't.

### PR DESCRIPTION
**Pull request by bcoca**
- updated non passing older test
  - switched to 'is' testing for status

fixes #39414 

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

playbooks

### CODE REVIEW

**Comment by sivel** [comment]
Related:
* https://github.com/ansible/ansible/issues/42352
* https://github.com/ansible/ansible/issues/34591

File: changelogs/fragments/conditionals_fix.yml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+bugfixes:
+    - remove bare var handling from conditionals (not needed since we removed bare vars from `with_` loops) to normalize hanlding of
```
Comment (decentral1se) [inline changelogs/fragments/conditionals_fix.yml]: `%s/hanlding/handling/`

File: lib/ansible/playbook/conditional.py [hunk 2]
```diff
@@ -113,21 +114,16 @@ def _check_conditional(self, conditional, templar, all_vars):
         if isinstance(conditional, bool):
             return conditional
 
+        if C.CONDITINAL_BARE_VARS:
+            if conditional in all_vars and VALID_VAR_REGEX.match(conditional):
+                display.deprecated('evaluating %s as a bare variable, this behaviour will go away and you might need to add |bool'
```
Comment (jamescassell) [inline lib/ansible/playbook/conditional.py:119]: Would be clearer to say we are automatically converting to bool.  "evaluating as a bare variable" didn't make sense to me and makes it sound like `when: enable_my_feature` is improper usage.  (My roles require the caller to pass bool's as such, not as strings.)

**Review by jamescassell — APPROVED** [review]
Overall, huge +1 to the change to make 'when: myvar' and 'when: not myvar' return opposite results.

File: lib/ansible/playbook/conditional.py (hunk 2 above)
Comment (bcoca) [reply lib/ansible/playbook/conditional.py:119]: but we are not automatically converting to a bool in this case, we are using the variable in 'python bool context' .. which might make it behave as if converting,  vs 'jinja2 if expression' which has slightly diff rules.

File: lib/ansible/config/base.yml [hunk 3]
```diff
@@ -318,6 +318,18 @@ COLOR_WARN:
   env: [{name: ANSIBLE_COLOR_WARN}]
   ini:
   - {key: warn, section: colors}
+CONDITINAL_BARE_VARS:
+  name: Allow bare variable evaluation in conditionals
+  default: True
+  type: boolean
+  description:
+    - With this setting on (True), ruring conditional evaluation 'var' is treated differently 'var.subkey' as the first is evaluted
```
Comment (maxamillion) [inline lib/ansible/config/base.yml]: is `ruring` a typo or a vocab term I just don't know?

**Comment by tgadiev** [comment]
Can we expect this PR to be merged soon?

**Comment by maxamillion** [comment]
rebuild_merge

**Comment by sivel** [comment]
bot_status

**Comment by bcoca** [comment]
bot_status

**Comment by tgadiev** [comment]
Related ansible-lint question:
 [#457](https://github.com/ansible/ansible-lint/issues/457)

**Comment by tgadiev** [comment]
@bcoca Can we expect backport PR for this fix against the stable-2.5, 2.6 & 2.7 branches?

**Comment by bcoca** [comment]
@tgadiev no, since the 'fix' is done by introducing a new feature. We don't backport new features.

**Comment by tgadiev** [comment]
> @tgadiev no, since the 'fix' is done by introducing a new feature. We don't backport new features.
@bcoca thanks for fast response!

Ok, It's understood for new features. But the bug  [#39414](https://github.com/ansible/ansible/issues/39414) is still persisting and affecting all currently supported versions from 2.5 to 2.7. If this fix is not compatible with old behavior does it make sense to provide another fix for older versions? We've made one - [#51029](https://github.com/ansible/ansible/pull/51029). It's compatible to old versions but needs some small fixes to be finished. Is it possible to reopen it and give a green light to that PR? We'll try to finish it if it's accepted for backporting to current stable versions. Currently this bug prevents many people to use correct and expected conditional syntax and blocks usage of current ansible-lint rules.

**Comment by bcoca** [comment]
We cannot fix previous versions w/o breaking existing playbooks, that is why the new toggle (feature) was introduced, this is also why backporting is a problem. 

The real 'fix' won't happen until a few versions go by and a) the default changes, b) most people switch to the new default and eventually we remove the toggle (feature) as it won't be needed. 

When dealing with things like this that affect existing usage at such a deep level we have opted for the approach above, in the past we just 'changed it' and had many people complain about their system just blowing up.

**Comment by tgadiev** [comment]
Well, currently ansible returns an exception and fails when trying to perform comparison of non-empty string top level variable like this:
```
when: my_var
```

If we change conditional behavior to prevent this failure and return True (as expected) but leave all the rest behavior - how it can break existing playbooks? It doesn't require to perform major change that breaks back compatibility. All current logic remains the same and shouldn't affect existing usage except removing bug that breaks playbook execution.

**Comment by tgadiev** [comment]
Please look at [#51029](https://github.com/ansible/ansible/pull/51029). This PR should fix the bug with string vars comparison but doesn't break current usage.

**Comment by tgadiev** [comment]
> @tgadiev it doesn't it handles one out of many case in which this happens, and not in a good way at that
> 
> i.e 'when: "cannot "

Ok. We can adjust it to handle it correctly. But the point is that we propose to adjust an exception handling to avoid breaking on simple comparison on non-empty string. This is workaround for old versions. It doesn't change the bare var handling totally like the given fix but allows people to prevent breaks on many cases like mentioned in [#39414](https://github.com/ansible/ansible/issues/39414). Is it possible to give us a chance to perform such fix?

**Comment by bcoca** [comment]
not sure what that woudl look like, but you are welcome to try

**Comment by tgadiev** [comment]
Ok. Thanks. So, can you re-open #51029 to let us try?

**Comment by bcoca** [comment]
i would submit new PR, there is little value in reopen

**Comment by tgadiev** [comment]
> i would submit new PR, there is little value in reopen

Ok. Shall we raise a new one instead of closed?

File: lib/ansible/config/base.yml [hunk 4]
```diff
@@ -318,6 +318,19 @@ COLOR_WARN:
   env: [{name: ANSIBLE_COLOR_WARN}]
   ini:
   - {key: warn, section: colors}
+CONDITINAL_BARE_VARS:
```
Comment (Glandos) [inline lib/ansible/config/base.yml:321]: Nobody noticed the typo here?

It lacks an O in CONDITIONAL. Tests should fail, or I missed something.

Comment (bcoca) [reply lib/ansible/config/base.yml:321]: its a typo, but it is consistently typoed on consumption also

Comment (Glandos) [reply lib/ansible/config/base.yml:321]: In `conditional.py`, yes, but the test runs the playbook with `ANSIBLE_CONDITIONAL_BARE_VARS`. Is it ok?

Comment (bcoca) [reply lib/ansible/config/base.yml:321]: why wouldn't it be?  That is the 'env variable' , which is not the thing that was misspelled (check 9 lines below)

Comment (bcoca) [reply lib/ansible/config/base.yml:321]: fixed in #53809

**Comment by kesor** [comment]
Does this mean that `else` and `endif` is not supported anymore?

```
 [WARNING]: conditional statements should not include jinja2 templating
delimiters such as {{ }} or {% %}. Found: {%- if aws_region is defined and
'cn-' in aws_region -%} True {%- else -%} False {%- endif -%}
```

**Comment by sivel** [comment]
@kesor your conditional was overly verbose before.  Ansible wraps what you provide to it implicitly in jinja2 already.

As such, what you want is:
```
when: aws_region is defined and 'cn-' in aws_region
```

**Pull request merged**